scripts/circuit_pk.py

```diff
@@ -0,0 +1,726 @@
+    '''
+
+    n = args.n
+    qis = args.qis
+    qis = json.loads(qis)
+    t = args.t
+
+    crt_moduli = CRTModuli(qis)
+    sigma = 3.2
+    discrete_gaussian = DiscreteGaussian(sigma)
+    bfv_crt = BFVCrt(crt_moduli, n, t, discrete_gaussian)
+
+    # Perform encryption of m in each CRT basis
+    s = bfv_crt.SecretKeyGen()
+    e = bfv_crt.bfv_q.rlwe.SampleFromErrorDistribution()
+    ais = []
+    
+    for i in range(len(crt_moduli.qis)):
+        ais.append(bfv_crt.bfv_qis[i].rlwe.Rq.sample_polynomial())
+    
+    # Generate Public key 
+    pub_key = bfv_crt.PublicKeyGen(s,e,ais)
+ 
+    
+    m = bfv_crt.bfv_q.rlwe.Rt.sample_polynomial()
+    e0 = bfv_crt.bfv_q.rlwe.SampleFromErrorDistribution()
+    e1 = bfv_crt.bfv_q.rlwe.SampleFromErrorDistribution()
+    u = bfv_crt.bfv_q.rlwe.SampleFromTernaryDistribution()
+
+    
+
+    ciphertext = bfv_crt.PubKeyEncrypt(pub_key,m,e0,e1,u)
+
+    # Sanity check for valid decryption    
+    message_prime = bfv_crt.Decrypt(s, ciphertext)
+
+    assert m == message_prime
+
+    # k1 = [QM]t namely the scaled message polynomial
+    k1 = m.scalar_mul(crt_moduli.q)
+    k1.reduce_coefficients_by_modulus(t)
+
+    # `p` is the modulus of the prime field of the circuit
+    p = 21888242871839275222246405745257275088548364400416034343698204186575808495617
+
+    # `r2is` are the polynomials r2i for each i-th CRT basis.
+    r2is = []
+
+    # `r1is` are the polynomials r1i for each i-th CRT basis.
+    r1is = []
+
+    # `k0is` are the negative multiplicative inverses of t modulo each qi.
+    k0is = []
+
+    # `ct0is` are the polynomials ct0i for each CRT basis. 
+    ct0is = []
+
+    # `ct0is_hat` are the polynomials ct0i_hat for each CRT basis.
+    ct0is_hat = []
+
+    # `ct1is` are the polynomials ct0i for each CRT basis. 
+    ct1is = []
+
+    # `ct1is_hat` are the polynomials ct0i_hat for each CRT basis.
+    ct1is_hat = []
+
+    # `pk0is` are the polynomials pk0 for each i-th CRT basis
+    pk0is = [] 
+
+    # `pk1is` are the polynomials pk0 for each i-th CRT basis
+    pk1is = []  
+
+    # `p1is` are the polynomials p1i for each i-th CRT basis.
+    p1is = []
+
+    # `p2is` are the polynomials p2i for each i-th CRT basis.
+    p2is = []
+
+
+
+    '''
+    SETUP PHASE - performed outside the circuit
+    For each CRT basis, we need to compute the polynomials r1i and r2i (check this doc for more details: https://hackmd.io/@gaussian/r1W98Kqqa)
+    '''
+    pk_array = []
+    pk1_array = []
+
+
+    for i,pk in enumerate(pub_key):
+        pk_array.append(pk[0].coefficients)
+        pk1_array.append(pk[1].coefficients)
+
+    
+        
+    cyclo = [1] + [0] * (n - 1) + [1]
+    cyclo = Polynomial(cyclo)
+
+    for i, cti in enumerate(ciphertext):
+
+        ct0i = cti[0]
+        ct1i = cti[1]
+
+        # k0i = -t^{-1} namely the multiplicative inverse of t modulo qi
+        k0i = pow(-t, -1, qis[i])
+
+        #ct0i_hat = pk0 *u + e0 + k0*k1
```

Comment:
Please in the in line comments describing formulas try to be consistent in the spacing. This should be 

There should always be a space after the initial `#`. 

```suggestion
        # ct0i_hat = pk0 * u + e0 + k0 * k1
```